### PR TITLE
add controller for cluster templates

### LINKED ISSUE #7319: Create controller for cluster template instances

### PR DESCRIPTION
**Pull request by zreigz**
**What this PR does / why we need it**: add a controller for cluster templates. This controller monitors cluster template instances and creates/deletes the user clusters depends on the number of replicas

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #7319

```release-note
Templates for cluster creation
```

### CODE REVIEW

**Review by mfranczy — APPROVED** [review]
/approve
/lgtm

File: pkg/controller/seed-controller-manager/cluster-template-controller/controller.go [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+*/
+
+package clustertemplatecontroller
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"k8s.io/apimachinery/pkg/types"
```
Comment (lsviben) [inline pkg/controller/seed-controller-manager/cluster-template-controller/controller.go]: NIT: sort imports

File: pkg/controller/seed-controller-manager/cluster-template-controller/controller.go [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+		if err := r.patchFinalizer(ctx, instance, true); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	// initialization
+	if !kuberneteshelper.HasFinalizer(instance, finalizer) {
+		if err := r.patchFinalizer(ctx, instance, false); err != nil {
```
Comment (lsviben) [inline pkg/controller/seed-controller-manager/cluster-template-controller/controller.go]: NIT: could you add constants like 
```
var remove = true
var add = false
```
so when ```r.patchFinalizer(ctx, instance, remove)``` is called its more clear

**Review by lsviben — COMMENTED** [review]
some minor comments, looks good

File: pkg/controller/seed-controller-manager/cluster-template-controller/controller.go (hunk 1 above)
Comment (zreigz) [reply pkg/controller/seed-controller-manager/cluster-template-controller/controller.go]: done

File: pkg/controller/seed-controller-manager/cluster-template-controller/controller.go (hunk 2 above)
Comment (zreigz) [reply pkg/controller/seed-controller-manager/cluster-template-controller/controller.go]: done

File: pkg/controller/seed-controller-manager/cluster-template-controller/controller.go (hunk 1 above)
Comment (aborilov) [reply pkg/controller/seed-controller-manager/cluster-template-controller/controller.go]: needs to be sorted again :)

Comment (zreigz) [reply pkg/controller/seed-controller-manager/cluster-template-controller/controller.go]: done

**Comment by aborilov** [comment]
👍 
/lgtm
/approve

**Pull request merged**